cli/EAP7-943_Security_Auth_Commands.adoc

```diff
@@ -192,26 +200,29 @@ _security disable-http-auth-management [--mechanism=<a string>]
  
 ==== enable-http-auth-http-server
 
-Associate an elytron HTTP factory to the security domain. In case a SASL factory already exists, 
+Associate an elytron HTTP factory to the security domain. In case an HTTP factory already exists, 
 the factory is extended(or updated) with the selected mechanism. 
 If no mechanism is provided, the Out of The Box Application HTTP factory is 
 associated to the security domain. This command creates all the non existing 
 resources required to configure authentication. This command can be called multiple 
 times to configure the referenced HTTP factory. Similar behaviour than _enable-http-auth-management_ command. +
  
-_security enable-http-auth-http-server --security-domain=<a string> (
-                                          [--file-system-realm=<a string>]
-                                          [--user-role-decoder=<a string>] |
-                                          [--user-properties-file=<null>]
-                                          [--group-properties-file=<null>]
-                                          [--relative-to=<a string>] )
-                                          [--mechanism=<a string>]
-                                          [--exposed-realm-name=<a string>]
-                                          [--external-trust-store=<a string>]
-                                          [--new-auth-factory-name=<a string>]
-                                          [--new-realm-name=<a string>]
-                                          [--new-security-domain-name=<a
-                                          string>] [--no-reload]_
+_security enable-http-auth-http-server --security-domain=<name> (
+                                          [--file-system-realm-name=<realm name>]
+                                          [--user-role-decoder=<role-decoder name>] |
+                                          [--user-properties-file=<file path>]
+                                          [--group-properties-file=<file path>]
+                                          [--relative-to=<system property>] |
+                                          [--properties-realm-name=<realm name>] | 
+                                          [--key-store-realm-name=<a string>] |
+                                          [--key-store-name=<key-store name>] )
+                                          [--mechanism=<mechanism name>]
```

Comment:
Yes, no way to specify roles. To be honest, I didn't think at it. Do you think tha tis a common use-case?

Reply:
I would say it is pretty common. Then other question is what to expose, probably just --group-properties-file will be enough.

But maybe we just hit the border - "High level commands can do less then native elytron commands".

In practice CLIENT_CERT (auth) + LDAP store (authz) is pretty common. For puspose of "hello world" scenarios CLIENT_CERT + property file will be enough.

Reply:
I have make it work without roles, but it still make me thinking that it is much less useful than possibility to specify roles as well.

Reply:
I looked at role + EXTERNAL SASL mechanism (I guess same for HTTP CLIENT_CERT) and it seems that we would need to:
1) allow to set an x500-principal-decoder if we want to map from a certificate to a user name
2) allow to provide the role-decoder (can be groups or roles properties file, could be a constant set of roles).
3) modify the SSL interactive support, it seems that the alias of the certificate inside the truststore needs to be exactly the same as what the X500 decoder decodes. Today we generate the alias, no way for a user to set it.

Would you agree with this? I would need to experiment with it to see the exact impact.

Reply:
Yes that will be the same.
1. +1
2. +1
3. +1 having alias option support seems useful in any case.

Reply:
Having the EXTERNAL mechanism + roles derived from properties file group brings a complexity, a principal-decoder must be set on the security-domain to map from the DN to an entry in the mgmt-groups.properties file. Unfortunately the DN can't be used in the file directly. I am not confortable in making changes to the security domain. This can impact all realms present in it. There doesn't seem to have ways to set the principal-decoder on the referenced realm inside the security domain. Perhaps @fjuma can comment on this. I would prefer not to deal with x500-principal-decoder, this implies the trust-store alias to be aligned with the decoder output.

Management groups are legacy, a constant role mapper could be more adequate in our case. We could propose to re-use an existing constant-role-mapper or provide a list of roles something like: --roles and --role-mapper

Reply:
Instead of x500-principal-decoder, another option might be to specify a principal-transformer on the referenced realm.

Reply:
@fjuma , thank-you, yes that works. I noticed that the user identity is not impacted by a transformer (the DN is kept), seems only used by the authentication layer. With a decoder, the identity is replaced by what the decoder returns. That is visible using whoami op.
One source of possible complexity is the fact that in both cases (decoder or transformer),  the returned "identity" must be a valid alias in the trust-store. Why do we have this limitation? For information I am using an aggregate-realm containing a key-store-realm and a properties file realm.

Reply:
One thing to note is that the principal-decoder resource also exposes the principal-transformer capability. This means that a principal-decoder can also be referenced as a principal-transformer on the mechanism-configuration and server-ssl-context - more details on this can be found in https://issues.jboss.org/browse/JBEAP-11084.

Reply:
@fjuma , so it seems that the way to go is to use a decoder as a realm transformer. In this case the user identity is what the decoder decodes. This seems what I would expect as a user when setting some kind of decoder.
I looked at https://issues.jboss.org/browse/JBEAP-11084 but don't see alias related information. I can see an HHTP/2 issue, but the 1 to 1 mapping between alias and decoder output is already taken into account as a constraint.

Reply:
To map between client principal to key-store alias I think you need post-realm-principal-transformer as described here http://darranl.blogspot.cz/2017/07/wildfly-elytron-principal-transformers.html
```
    The Principal created by the pre-realm-principal-transformers is: -
        The Principal used to map the SecurityRealm
        The Principal that will be associated with the resulting SecurityIdentity.
    The Principal created after the final principal transformers is: -
        The Principal that will be passed to the SecurityRealm to obtain the RealmIdentity.
```

Reply:
I guess that the problem comes from the aggregate-realm. I am setting the principal-transformer on it so the authentication realm is impacted by it and requires a valid alias in the trust-store. I should be able to set a principal-decoder for the authorization realm only. @fjuma does it make sense?

Reply:
FYI I have logged https://issues.jboss.org/browse/WFCORE-3767 that would separate alias and principal in this case. For now anyway we are relying on a constant set of roles.